File: core/src/test/java/dev/morphia/critter/parser/TestPropertyModelGenerator.java
package dev.morphia.critter.parser;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

import dev.morphia.critter.CritterClassLoader;
import dev.morphia.mapping.codec.pojo.EntityModel;
import dev.morphia.mapping.codec.pojo.PropertyModel;
import dev.morphia.mapping.codec.pojo.critter.CritterPropertyModel;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class TestPropertyModelGenerator extends BaseCritterTest {
    private final CritterClassLoader critterClassLoader = new CritterClassLoader();

    // @ParameterizedTest
    @MethodSource("properties")
    @DisplayName("")
    public void testProperty(String control, String methodName, Method method) throws Exception {
        CritterPropertyModel propertyModel = getModel(control);
        System.out.println("exampleModel = [" + control + "], methodName = [" + methodName + "], method = [" + method + "]");
        Object expected = method.invoke(control);
        Object actual = method.invoke(propertyModel);
        Assertions.assertEquals(expected, actual, method.getName() + " should return the same value");
    }

    private CritterPropertyModel getModel(String name) {
        return (CritterPropertyModel) GeneratorTest.entityModel.getProperty(name);
    }

    static Stream<Arguments> properties() {
        Object[][] methods = GeneratorTest.methodNames(CritterPropertyModel.class);
        return List.of("dev.morphia.critter.sources.ExampleNamePropertyModelTemplate").stream()
                .map(type -> {
                    try {
                        return loadModel(type);
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                })
                .flatMap(propertyModel -> Arrays.stream(methods)
                        .map(method -> Arguments.of(propertyModel.getName(), method[0], method[1])));
    }

    private static PropertyModel loadModel(String type) throws Exception {
        return (PropertyModel) new CritterClassLoader()
                .loadClass(type)
                .getConstructor(EntityModel.class)
                .newInstance(GeneratorTest.entityModel);
    }
}

File: critter/core/src/test/kotlin/dev/morphia/critter/parser/TestPropertyModelGenerator.kt
package dev.morphia.critter.parser

import dev.morphia.critter.Critter.Companion.critterClassLoader
import dev.morphia.critter.parser.GeneratorTest.entityModel
import dev.morphia.critter.parser.GeneratorTest.methodNames
import dev.morphia.mapping.codec.pojo.EntityModel
import dev.morphia.mapping.codec.pojo.PropertyModel
import dev.morphia.mapping.codec.pojo.critter.CritterPropertyModel
import java.lang.reflect.Method
import org.testng.Assert.assertEquals
import org.testng.annotations.DataProvider
import org.testng.annotations.NoInjection

class TestPropertyModelGenerator : BaseCritterTest() {
    //    @Test(dataProvider = "properties", testName = "")
    fun testProperty(control: String, methodName: String, @NoInjection method: Method) {
        val propertyModel = getModel(control)

        println("exampleModel = [${control}], methodName = [${methodName}], method = [${method}]")
        val expected = method.invoke(control)
        val actual = method.invoke(propertyModel)
        assertEquals(actual, expected, "${method.name} should return the same value")
    }

    private fun getModel(name: String): CritterPropertyModel {
        return entityModel.getProperty(name) as CritterPropertyModel
    }

    @DataProvider(name = "properties")
    fun methods(): Array<Array<Any>> {
        val methods = methodNames(CritterPropertyModel::class.java)
        return listOf("dev.morphia.critter.sources.ExampleNamePropertyModelTemplate")
            .map { loadModel(it) }
            .flatMap { propertyModel ->
                methods.map { method -> arrayOf(propertyModel.name, method[0], method[1]) }
            }
            .toTypedArray()
    }

    private fun loadModel(type: String): PropertyModel {
        return critterClassLoader
            .loadClass(type)
            .getConstructor(EntityModel::class.java)
            .newInstance(entityModel) as PropertyModel
    }
}
